8305637: Remove Opaque1 nodes for Parse Predicates and clean up useless predicate elimination

chhagedorn opened the pull request:
This is the last clean-up PR before the complete fix for Assertion Predicates ([JDK-8288981](https://bugs.openjdk.org/browse/JDK-8288981)). 

This patch includes:
- Removal of `ConI`->`Opaque1`->`Conv2B` input nodes for `ParsePredicateNodes` with the following additional changes:
  - Adjusting `ParsePredicateNode` to block unwanted optimizations (added empty `ParsePredicateNode::Ideal()`).
  - Changing `Compile::_parse_predicate_opaqs` to not store `Opaque1Nodes` to keep track of Parse Predicates but instead storing `ParsePredicateNodes` directly. Renamed to `Compile::_parse_predicates` and adjusted related methods.
  - Removed asserts matching `Opaque1` -> `Conv2B` shape.
- Cleaning up `eliminate_useless_predicates()`:
  - Adjust code to find useful/useless Parse Predicates with the new `Compile::_parse_predicates` list with `ParsePredicateNodes` instead of `Opaque1Nodes`.
  - Changing `ParsePredicateNode` to carry a `_useless` state which simplifies the elimination of useless predicates with `eliminate_useless_predicates()` and during IGVN (added `ParsePredicateNode::Value()` for that which also removes the predicate once we are in post loop opts IGVN). 
- Some refactoring/clean-ups of involved code.
  
Testing: tier1-7 + some fuzzer testing

Thanks,
Christian

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305637](https://bugs.openjdk.org/browse/JDK-8305637): Remove Opaque1 nodes for Parse Predicates and clean up useless predicate elimination (**Sub-task** - P3)


### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [763a6b4a](https://git.openjdk.org/jdk/pull/15449/files/763a6b4a40188d942cda23469da176c86f4d6a7d)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15449/head:pull/15449` \
`$ git checkout pull/15449`

Update a local copy of the PR: \
`$ git checkout pull/15449` \
`$ git pull https://git.openjdk.org/jdk.git pull/15449/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15449`

View PR using the GUI difftool: \
`$ git pr show -t 15449`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15449.diff">https://git.openjdk.org/jdk/pull/15449.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15449#issuecomment-1695832476)

rwestrel reviewed (commented):
Opaque1 nodes fold away after loop opts which guarantees the parse predicate are removed too after loop opts. In the new code, without the Opaque1 nodes, what causes the parse predicate to be removed after loop opts?

chhagedorn commented:
> Opaque1 nodes fold away after loop opts which guarantees the parse predicate are removed too after loop opts. In the new code, without the Opaque1 nodes, what causes the parse predicate to be removed after loop opts?

When creating a new `ParsePredicateNode`, we are registering it for post loop opts IGVN:
https://github.com/openjdk/jdk/blob/d50356e14877b4d4376fc18ecdec29f7d98d77bd/src/hotspot/share/opto/ifnode.cpp#L1978-L1984

During post loop opts IGVN (or earlier if `_useless` is true), the `ParsePredicateNode` is folded when calling `ParsePredicateNode::Value()`:
https://github.com/openjdk/jdk/blob/d50356e14877b4d4376fc18ecdec29f7d98d77bd/src/hotspot/share/opto/ifnode.cpp#L2004-L2011

I don't think that _useful_ `ParsePredicateNodes` should ever be cloned. If a loop is removed, then either major progress is true and `eliminate_useless_predicates()` will mark them useless (i.e. set `_useless` to true) in the next round of loop opts or if major progress was false then `Compile::mark_parse_predicate_nodes_useless()` have already marked them useless. In both cases, they will be removed by the next round of IGVN. Otherwise, I don't think that useful `ParsePredicateNodes` should ever be split (i.e. being part of a loop body to be cloned) or cloned otherwise (i.e. by split-if etc.). When cloning them to unswitched loops, I create them newly by calling the constructor, so they end up on the post loop opts IGVN list:
https://github.com/openjdk/jdk/blob/d50356e14877b4d4376fc18ecdec29f7d98d77bd/src/hotspot/share/opto/loopPredicate.cpp#L310-L311

However, there are some cases where useless `ParsePredicateNodes` are cloned. For example, parsing could have added them to Ifs inside a loop. But then `eliminate_useless_predicates()` will mark them useless. When splitting the loop and cloning these `ParsePredicateNodes`, they will still be marked useless and they are all cleaned up during the next round of IGVN. So, I think registering `ParsePredicateNodes` for post loop opts once in the constructor should be enough.

Nevertheless, I still register them in the `Compile::_parse_predicates` list in `Node:clone()` to not miss this:
https://github.com/openjdk/jdk/blob/d50356e14877b4d4376fc18ecdec29f7d98d77bd/src/hotspot/share/opto/node.cpp#L511-L513

After post loop opts IGVN, I assert that we got rid of all the `ParsePredicateNodes`:
https://github.com/openjdk/jdk/blob/d50356e14877b4d4376fc18ecdec29f7d98d77bd/src/hotspot/share/opto/compile.cpp#L1856-L1858

rwestrel commented:
> > Opaque1 nodes fold away after loop opts which guarantees the parse predicate are removed too after loop opts. In the new code, without the Opaque1 nodes, what causes the parse predicate to be removed after loop opts?
> 
> When creating a new `ParsePredicateNode`, we are registering it for post loop opts IGVN:

Makes sense. Thanks for the details.

rwestrel commented on src/hotspot/share/opto/multnode.cpp:
```diff
@@ -222,19 +222,15 @@ CallStaticJavaNode* ProjNode::is_uncommon_trap_if_pattern(Deoptimization::DeoptR
-      return nullptr;
-    }
+  // we need a ParsePredicate node for predicate reasons
+  if (reason != Deoptimization::Reason_none && !iff->is_ParsePredicate()) {
```
Unrelated to your change but this code doesn't seem to do what the comment says. `reason != Deoptimization::Reason_none` is not "predicate reasons". I think this needs to be cleaned up at some point.

chhagedorn replied:
Okay, I've missed one usage which is not `Deoptimization::Reason` - but let me still have a look if I can clean this method up somehow.

rwestrel reviewed (approved):
Looks good to me.

chhagedorn commented:
I've pushed an update with the following changes:
- `is_uncommon_trap_if_pattern()` and `is_uncommon_trap_proj()`:
  - Changed `Deoptimization::Deopt_Reason` into a default argument for `is_uncommon_trap_if_pattern()` and `is_uncommon_trap_proj()` as most calls are only interested if the `If` just represents such a pattern. This allowed us to remove all the `Deoptimization::Reason_none` arguments in the calling code.
  - Changed them into `const` methods.
  - Removed/Moved Parse Predicate checking in `is_uncommon_trap_if_pattern()` as this should be the responsibility of the caller. The method simply wants to check if there is such a pattern with the given deopt reason.
- `create_new_if_for_predicate()`:
  - The continuation projection is always a Parse Predicate success projection (i.e. true projection). Changed types and variable names accordingly.
  - Other renaming and removing dead code.
- Changed some variable names to better highlight that where Parse Predicate success projections are passed around.

But I could also split these changes in a separate PR.

chhagedorn commented:
Thanks Roland for your review!

vnkozlov reviewed (approved):
Looks reasonable.

chhagedorn commented:
Thanks Vladimir for your review! 

I'm currently running some more testing. Will integrate it tomorrow if it looks good.

chhagedorn commented:
Testing looked good.

/integrate
